R/tooltip.R

```diff
@@ -0,0 +1,125 @@
+#'   assistive technology users.
+#' @param ... UI elements for the tooltip. Character strings are automatically
+#'   [htmlEscape()]d unless marked as [HTML()].
+#' @param id If provided, you can use `input$id` in your server logic to
+#'   determine whether the tooltip is currently shown/hidden.
+#' @param placement The placement of the tooltip relative to its trigger.
+#' @param options A list of additional [Bootstrap
+#'   options](https://getbootstrap.com/docs/5.3/components/tooltips/#options).
+#'
+#' @details If `trigger` yields multiple HTML elements (i.e., it's a `tagList()`
+#'   or complex `{htmlwidgets}` object), the last HTML element is used as the
+#'   tooltip's target. If the `trigger` should contain all of those elements,
+#'   consider wrapping the object in a [span()] or [div()].
+#'
+#' @describeIn tooltip Add a tooltip to a UI element
+#' @references <https://getbootstrap.com/docs/5.3/components/tooltips/>
+#' @export
+#' @examplesIf interactive()
+#'
+#' tooltip(
+#'   shiny::actionButton("btn", "A button"),
+#'   "A message"
+#' )
+#'
+#' card(
+#'   card_header(
+#'     tooltip(
+#'       span("Card title ", bsicons::bs_icon("question-circle-fill")),
+#'       "Additional info",
+#'       placement = "right"
+#'     )
+#'   ),
+#'   "Card body content..."
+#' )
+tooltip <- function(
+  trigger,
+  ...,
+  id = NULL,
+  placement = c("auto", "top", "right", "bottom", "left"),
+  options = list()
+) {
+
+  args <- rlang::list2(...)
+  argnames <- rlang::names2(args)
+
+  children <- args[!nzchar(argnames)]
+  attribs <- args[nzchar(argnames)]
+
+  if (length(children) == 0) {
+    abort("At least one value must be provided to `...`.")
+  }
+
+  res <- web_component(
+    "bslib-tooltip",
+    id = id,
+    placement = rlang::arg_match(placement),
+    options = jsonlite::toJSON(options, auto_unbox = TRUE),
+    !!!attribs,
+    # Use display:none instead of <template> since shiny.js
+    # doesn't bind to the contents of the latter
+    div(!!!children, style = "display:none;"),
+    trigger
+  )
+
+  res <- tag_require(res, version = 5, caller = "tooltip()")
+  as_fragment(res)
+}
+
+#' @describeIn tooltip Programmatically show/hide a tooltip.
+#'
+#' @param id a character string that matches an existing tooltip id.
+#' @param show Whether to show (`TRUE`) or hide (`FALSE`) the tooltip. The
+#'   default (`NULL`) will show if currently hidden and hide if currently shown.
+#' @param session A Shiny session object (the default should almost always be
+#'   used).
+#'
+#' @export
+tooltip_toggle <- function(id, show = NULL, session = get_current_session()) {
+  show <- normalize_show_value(show)
+
+  msg <- list(method = "toggle", value = show)
+  force(id)
+  callback <- function() {
+    session$sendInputMessage(id, msg)
+  }
+  session$onFlush(callback, once = TRUE)
+}
+
+
+#' @describeIn tooltip Update the contents of a tooltip.
+#' @export
+tooltip_update <- function(id, ..., session = get_current_session()) {
```

Comment:
Should we support updating other aspects of the tooltip, like `placement` or `options`?

Reply:
Potentially, maybe in the future?